src/scf/direct_minimization.jl

```diff
@@ -66,20 +67,20 @@ function direct_minimization(basis::PlaneWaveBasis{T}, Psi0;
                              optim_solver=Optim.LBFGS, kwargs...) where T
     model = basis.model
     @assert model.spin_polarisation in (:none, :spinless)
-    @assert model.assume_band_gap
+    @assert model.assume_band_gap # temperature is not yet supported
```

Comment:
Add `@assert model.smearing === nothing` ?

Reply:
I'm not too clear on how that works. Why do we have a `assume_band_gap` as kwarg to Model? Shouldn't `assume_band_gap` just be `temperature==0`?

Reply:
The idea here is that one could have a system with a band gap and still employ temperature (!= smearing, but "real" temperature). Right now our setup breaks with this (see https://github.com/JuliaMolSim/DFTK.jl/blob/master/src/occupation.jl#L104), but this catches that.

Reply:
We should do it at the level of the model. What's a band gap with temperature? We should either do zero temperature or finite temperature, not sure what else there is?

Reply:
Physical finite temperature versus ficticious finite temperature to improve convergence in metals?

Reply:
Computationally there's no difference so why make one?

Reply:
Because it's different physics? I don't know, maybe we shouldn't make any?

Reply:
So what's going on now is a bit weird. The band gap thing switches between the two functions we have in occupation.jl, but the first one is sort of redundant : it's a particular case of the second one when there's a gap. The rational thing to do here is remove any mention of bandgap, and use the second function all the time, even at 0K. However I kinda like the first function because it's simpler and does not need a bisection. Maybe we disable the codepaths that lead to and mark it visually as inactive but leave it in the source and still test it?

Reply:
Anyway, that'll go in a separate PR. I'll open an issue

Reply:
You say weird, made perfect sense to me, when I coded it ;). I think the current structure was made, because you mentioned bisection would converge slowly if a system is gapped. Why would that be different now?

Assuming that `find_occupation_fermi_metal` is ok for gapped systems, the other functions are really redundant, especially branch `https://github.com/JuliaMolSim/DFTK.jl/blob/master/src/occupation.jl#L105`. I still do not see why we should not have an `assume_band_gap` flag in the model, which would allow to play some tricks e.g. in the occupation for the 0K case. Probably the logic for setting it should be simplified, that I agree (like `assume_band_gap = false` by default and `true` needs to be set explicitly).

Reply:
AFAICT the current structure is because we did bandgapped systems before fractional occupations, but it's kinda obsolete now. 

This is just an optimization, so it should be done at the innermost level. Bisection is pretty fast, and I don't really see a case where it should be the bottleneck (right now it should be pretty unoptimized because of type stability issues, but it never showed up on my profiling) If it turns out to be (in very specific cases, like ridiculously low Ecut with a lot of kpoints maybe? Even then I'm skeptical) we can optimize it later (it's easy enough to do, just compute the Fermi level as if it was bandgapped, and use that as starting point for bisection). I'll make a PR with that sometime after the incoming hyperglycemic coma. Merry Christmas!

Reply:
Hohoho. Same to you ;)